src/components/ActionButton/ActionButton.tsx

```diff
@@ -86,13 +94,24 @@ const ActionButton = ({
     }
 
     if (!loading && showLoader) {
+      const now = new Date();
+      // calculate elasped time of loading from user and do subtraction to LOADER_MIN_DURATION,
+      // also add an edge case when time diff is less than 0 to be 0.
+      const loadingMilliseconds: number =
+        now.getTime() - (startLoadTime.current ?? now).getTime();
+      loadingTimeTracker.current = Math.max(
+        LOADER_MIN_DURATION - loadingMilliseconds,
+        0,
+      );
       loaderTimeout = window.setTimeout(() => {
+        loadingTimeTracker.current = 0;
+        startLoadTime.current = undefined;
         setShowLoader(false);
 
         if (success) {
           setShowSuccess(true);
         }
-      }, LOADER_MIN_DURATION);
+      }, loadingTimeTracker.current);
```

Comment:
Do we need to store `loadingTimeTracker` as a ref? Can't we simply put here the result of the calculation above? Something like:
```suggestion
      }, Math.max(LOADER_MIN_DURATION - loadingMilliseconds, 0));
```
Then of course we also wouldn't need to reset it to zero inside the timeout :)

Reply:
And maybe skip the timeout if the value is zero:
```
      const loadFinishHandler = () => {
        startLoadTime.current = undefined;
        setShowLoader(false);

        if (success) {
          setShowSuccess(true);
        }
      };

      const loadingTimeout = Math.max(LOADER_MIN_DURATION - loadingMilliseconds, 0);
      if (loadingTimeout > 0) {
        setTimeout(loadFinishHandler, loadingTimeout);
      } else {
        loadFinishHandler();
      }
```